src/rnnlm/arpa-sampling.cc

```diff
@@ -0,0 +1,145 @@
+  int32 cur_order = ngram.words.size();
+  int32 word = ngram.words.back(); // word is the last word in a ngram term 
+  HistType history(ngram.words.begin(), ngram.words.begin() + cur_order - 1);
+  KALDI_ASSERT(history.size() == cur_order - 1);
+
+  // log probability of a ngram term (natural log)
+  BaseFloat log_prob = ngram.logprob;
+  // backoff log probability of a ngram term (natural log)
+  BaseFloat backoff_weight = ngram.backoff;
+  std::pair <BaseFloat, BaseFloat> probs_pair;
+  probs_pair = std::make_pair(log_prob, backoff_weight);
+  probs_[cur_order - 1][history].insert({word, probs_pair});
+ 
+  // get the total number of words from an ARPA file
+  if (cur_order == 1) {
+    num_words_++;
+  }
+}
+
+void ArpaSampling::HeaderAvailable() {
+  ngram_counts_ = NgramCounts();
+  ngram_order_ = NgramCounts().size(); 
+  probs_.resize(ngram_order_);
+}
+
+BaseFloat ArpaSampling::GetProb(int32 order, int32 word, const HistType &history) {
+  BaseFloat prob = 0.0;
+  NgramType::const_iterator it = probs_[order - 1].find(history);
+  if (it != probs_[order - 1].end() &&
+      probs_[order-1][history].find(word) != probs_[order-1][history].end()) {
+    prob += probs_[order-1][history][word].first;
+  } else { // backoff to the previous order
+    order--;
+    if (order >= 1) {
+      HistType::const_iterator first = history.begin() + 1;
+      HistType::const_iterator last = history.end();
+      HistType h(first, last);
+      prob += GetProb(order, word, h);
+      int32 word_new = history.back();
+      HistType::const_iterator last_new = history.end() - 1;
+      HistType h_new(history.begin(), last_new);
+      prob += GetBackoffWeight(order, word_new, h_new);
+    }
+  }
+  return prob;
+}
+
+BaseFloat ArpaSampling::GetBackoffWeight(int32 order, int32 word, 
+    const HistType &history) {
+  BaseFloat bow = 0.0;
+  KALDI_ASSERT(order >= 1);
+  NgramType::const_iterator it = probs_[order - 1].find(history);
+  if (it != probs_[order - 1].end()) {
+    WordToProbsMap::const_iterator it2 = probs_[order - 1][history].find(word);
+    if (it2 != probs_[order - 1][history].end()) {
+      bow = it2->second.second;
+    }
+  }
+  return bow;
+}
+
+void ArpaSampling::GetUnigramDistribution(std::vector<BaseFloat> *unigram_probs) {
+  (*unigram_probs).clear();
+  (*unigram_probs).resize(num_words_ + 1);
+  for (int32 i = 0; i < num_words_ + 1; ++i) {
+    HistType h; // empty history
+    WordToProbsMap::const_iterator it = probs_[0][h].find(i);
+    if (it != probs_[0][h].end()) {
+      (*unigram_probs)[i] = Exp(it->second.first);
+    }    
+  }
+}
+
+void ArpaSampling::ComputeHistoriesWeights(
+    const std::vector<std::pair<HistType, BaseFloat> > &histories,
+    HistWeightsType *hists_weights) {
+  (*hists_weights).clear();
+  for (std::vector<std::pair<HistType, BaseFloat> >::const_iterator
+      it = histories.begin(); it != histories.end(); ++it) {
+    HistType history((*it).first);
```

Comment:
This copies the whole array to a newly allocated memory. And you are not modifying this array! There is no need to keep a copy. This is what const references are for:  
`const HistType& history = it->first;`